File: Nearest Neighbour Algorithm/nearest_neighbour_algorithm.js
const NUMBER_VERTICES = 100;
const RAD = 5

let vertices = [];
let visited_vertices = [0]; // By default, the graph starts with the first vertex
let current_vertex = 0;

function setup() {
  createCanvas(windowWidth, windowHeight);
  background(255);
  strokeWeight(RAD / 2);
  textSize(RAD * 2);
  for (let i = 0; i < NUMBER_VERTICES; i++) {
    vertices[i] = new Vertex(random(width), random(height));
    vertices[i].display();
    text(i, vertices[i].x + RAD * 2, vertices[i].y + RAD * 2);
  }
}

function draw() {
  let record_distance = Infinity;
  let nearest_vertex;
  for (let i = 0; i < vertices.length; i++) {
    if (!visited_vertices.includes(i)) {
      let d = dist(vertices[i].x, vertices[i].y, vertices[current_vertex].x, vertices[current_vertex].y);
      if (d < record_distance) {
        nearest_vertex = i;
        record_distance = d;
      }
    }
  }
  line(vertices[nearest_vertex].x, vertices[nearest_vertex].y, vertices[current_vertex].x, vertices[current_vertex].y);
  visited_vertices.push(nearest_vertex);
  current_vertex = nearest_vertex;
  
  if (visited_vertices.length == vertices.length) {
    noLoop();
  }
}

class Vertex {
  constructor(x_, y_) {
    this.x = x_;
    this.y = y_;
  }

  display() {
    fill(0);
    circle(this.x, this.y, RAD * 2);
  }
}

File: Nearest Neighbour Algorithm/Python/nearest_neighbour_algorithm.py
from math import sqrt
from random import randint


class Vertex:
    def __init__(self, x, y):
        self.x = x
        self.y = y


def dist(x1, y1, x2, y2):
    return sqrt((x1 - x2) ** 2 + (y1 - y2) ** 2)


NUMBER_VERTICES = 10
WIDTH = HEIGHT = 100  # dimension of the canvas

vertices = []
visited_vertices = [0]  # by default, the graph starts with the first vertex
path = [0]
current_vertex = 0

print("* Nearest Neighbour Algorithm *")
print("Number of vertices :", NUMBER_VERTICES, "| Dimensions of the canvas : (" + str(WIDTH), ";", str(HEIGHT) + ")\n")
print("Vertices coordinates :")
for i in range(NUMBER_VERTICES):
    vertices.append(Vertex(randint(0, WIDTH), randint(0, HEIGHT)))
    print(i, ": (" + str(vertices[i].x), ";", str(vertices[i].y) + ")")

while len(visited_vertices) < len(vertices):
    global nearest_vertex
    record_distance = dist(0, 0, WIDTH, HEIGHT)
    for i in range(len(vertices)):
        if i not in visited_vertices:
            d = dist(vertices[i].x, vertices[i].y, vertices[current_vertex].x, vertices[current_vertex].y)
            if d < record_distance:
                nearest_vertex = i
                record_distance = d
    visited_vertices.append(nearest_vertex)
    path.append(nearest_vertex)
    current_vertex = nearest_vertex

print("Path :", path)
print("Adjacency matrix :")
adjacency_matrix = [[0 for col in range(NUMBER_VERTICES)] for row in range(NUMBER_VERTICES)]
for i in range(NUMBER_VERTICES - 1):
    adjacency_matrix[path[i]][path[i + 1]] = 1
    adjacency_matrix[path[i + 1]][path[i]] = 1
for row in adjacency_matrix:
    print(*row)
    
